sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -469,6 +545,145 @@ object GpuHashAggregateIterator extends Logging {
+    // - Partial mode: grouping key + cudf aggregates (e.g. no avg, intead sum::count
+    // - Final mode: grouping key + spark aggregates (e.g. avg)
+    val finalAttributes = groupingAttributes ++ aggregateAttributes
+
+    // boundResultReferences is used to project the aggregated input batch(es) for the result.
+    // - Partial mode: it's a pass through. We take whatever was aggregated and let it come
+    //   out of the node as is.
+    // - Final or Complete mode: we use resultExpressions to pick out the correct columns that
+    //   finalReferences has pre-processed for us
+    val boundResultReferences = if (modeInfo.hasPartialMode || modeInfo.hasPartialMergeMode) {
+      GpuBindReferences.bindGpuReferences(
+        resultExpressions,
+        resultExpressions.map(_.toAttribute))
+    } else if (modeInfo.hasFinalMode || modeInfo.hasCompleteMode) {
+      GpuBindReferences.bindGpuReferences(
+        resultExpressions,
+        finalAttributes)
+    } else {
+      GpuBindReferences.bindGpuReferences(
+        resultExpressions,
+        groupingAttributes)
+    }
+    BoundExpressionsModeAggregates(
+      boundFinalProjections,
+      boundResultReferences)
+  }
+
+  private[this] def reorderFinalBatch(finalBatch: ColumnarBatch,
+      boundExpressions: BoundExpressionsModeAggregates,
+      metrics: GpuHashAggregateMetrics): ColumnarBatch = {
+    // Perform the last project to get the correct shape that Spark expects. Note this may
+    // add things like literals that were not part of the aggregate into the batch.
+    closeOnExcept(GpuProjectExec.projectAndClose(finalBatch,
+      boundExpressions.boundResultReferences, NoopMetric)) { ret =>
+      metrics.numOutputRows += ret.numRows()
+      metrics.numOutputBatches += 1
+      ret
+    }
+  }
+
+  def makeIter(cbIter: Iterator[ColumnarBatch],
+      boundExpressions: BoundExpressionsModeAggregates,
+      metrics: GpuHashAggregateMetrics): Iterator[ColumnarBatch] = {
+    val aggTime = metrics.computeAggTime
+    val opTime = metrics.opTime
+    cbIter.map { batch =>
+      withResource(new NvtxWithMetrics("finalize agg", NvtxColor.DARK_GREEN, aggTime,
+        opTime)) { _ =>
+        val finalBatch = boundExpressions.boundFinalProjections.map { exprs =>
+          GpuProjectExec.projectAndClose(batch, exprs, NoopMetric)
+        }.getOrElse(batch)
+        reorderFinalBatch(finalBatch, boundExpressions, metrics)
+      }
+    }
+  }
+
+  def makeIterFromSpillable(sbIter: Iterator[SpillableColumnarBatch],
+      boundExpressions: BoundExpressionsModeAggregates,
+      metrics: GpuHashAggregateMetrics): Iterator[ColumnarBatch] = {
+    val aggTime = metrics.computeAggTime
+    val opTime = metrics.opTime
+    sbIter.map { sb =>
+      withResource(new NvtxWithMetrics("finalize agg", NvtxColor.DARK_GREEN, aggTime,
+        opTime)) { _ =>
+        val finalBatch = boundExpressions.boundFinalProjections.map { exprs =>
```

Comment:
nit, since this is the final step for a partial (pre shuffle), should we call it `batchToShuffle`? To distinguish it from the batch that is merged and ready for a parent to process after a shuffle.

A comment around this function would help call out this is to be used in the single-pass agg case.

Reply:
But it is not guaranteed to be that way. This same code is also used for a complete or final aggregation. This is the "final" step of any of the aggregations. Should I rename it so that it is a last pass instead of a final pass so that the terms don't conflict with the aggregation types?

Reply:
And if I do, then I would want to rename boundFinalProjections right?